### PR TITLE
feat: verify contracts

### PR DESCRIPTION
**Pull request by Kittikornl**
- add verifier function in core
- add verifier class for each module
- note:
 1.Testing needs to be done manually since the contract can be verified one time only
 2.support only etherscan

### CODE REVIEW

File: packages/modules/token/src/utils/verifier/index.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    apiKey?: string
+    apiUrl: string
+    url: string
+}
+
+export class TokenVerifier {
+  etherscan: Etherscan
+  client: PublicClient
+    constructor(chain: Chain, rpcEndpoint: string, params: BlockExplorerParams) {
+        params.apiKey = params.apiKey ?? "API_KEY_NOT_PROVIDED" // note: test that we can use default value
```
Comment (tapakornl) [inline packages/modules/token/src/utils/verifier/index.ts:18]: why do we need to assgin `API_KEY_NOT_PROVIDED` to `params.apiKey`, can't we just assign `""`

File: packages/modules/aave-v3/contracts/aave-v3-core/hardhat.config.cjs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 const { baseHardhatUserConfig } = require('@infinit-xyz/core/internal/hardhat-base-config')
 const { name } = require('../../package.json')
+require('@nomicfoundation/hardhat-verify')
```
Comment (melvinmcrn) [inline packages/modules/aave-v3/contracts/aave-v3-core/hardhat.config.cjs:3]: @Kittikornl Do we need to import hardhat-verify? I thought the config is only used when compile?

File: packages/modules/aave-v3/contracts/aave-v3-periphery/hardhat.config.cjs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 const { baseHardhatUserConfig } = require('@infinit-xyz/core/internal/hardhat-base-config')
 const { name } = require('../../package.json')
+require('@nomicfoundation/hardhat-verify')
```
Comment (melvinmcrn) [inline packages/modules/aave-v3/contracts/aave-v3-periphery/hardhat.config.cjs:3]: same as above comment

File: packages/modules/aave-v3/src/utils/contract-verifier.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    return contracts
+  }
+
+  private async getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies: Record<string, Address>) {
+    const contracts: ContractInfo[] = []
+    for (const address of Object.values(reserveInterestRateStrategies)) {
+      contracts.push({ address })
+    }
+    return contracts
+  }
```
Comment (melvinmcrn) [inline packages/modules/aave-v3/src/utils/contract-verifier.ts:85]: please add test for these methods

File: packages/core/src/utils/contract-verifier/helper/index.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export { verifyContract } from './verifyContract'
```
Comment (melvinmcrn) [inline packages/core/src/utils/contract-verifier/helper/index.ts:1]: @Kittikornl  please add test for each helper method

File: packages/core/src/utils/contract-verifier/baseContractVerifier.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
+import { PublicClient } from 'viem'
+
+import { Artifacts } from 'hardhat/types'
+
+import { Etherscan } from '@nomicfoundation/hardhat-verify/etherscan.js'
+import { verifyContract } from './helper'
+import { BlockExplorerParams, ContractInfo, ContractVerifierCallback } from './type'
```
Comment (tapakornl) [inline packages/core/src/utils/contract-verifier/baseContractVerifier.ts:7]: import absolute path

File: packages/core/src/utils/contract-verifier/baseContractVerifier.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+   */
+  constructor(client: PublicClient, params: BlockExplorerParams) {
+    // NOTE: 
+    // 1. We can use Etherscan class for Blockscout API as well.
+    // 2. Blockscout API key is not required. but Etherscan client requires it.
+    // so we use a default value if it is not provided.
+    // 3. We only support Blockscout v1 API.
+    const parsedApiKey = params.apiKey ?? 'BLOCKSCOUT'
+    // trim the /v2 from the end of the URL
+    params.apiUrl = params.apiUrl.split('/v2')[0]
```
Comment (tapakornl) [inline packages/core/src/utils/contract-verifier/baseContractVerifier.ts:32]: if we don't support api v2, I think we should throw error

**Review by melvinmcrn — DISMISSED** [review]

File: packages/core/src/utils/contract-verifier/helper/__tests__/getContractCreationCode.integration.test.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { describe, expect, test } from 'vitest'
+
+import { getContractCreationCode } from '../getContractCreationCode'
+import { Etherscan } from '@nomicfoundation/hardhat-verify/etherscan'
+
+describe('getCreationCodeFromBlockExplorer', () => {
+  const apiKey = 'U19FHTSVAE2MTBTG89IWKF6ERWW9VX9IY4'
```
Comment (mewpawee) [inline packages/core/src/utils/contract-verifier/helper/__tests__/getContractCreationCode.integration.test.ts:7]: should we set as a env variable?

File: packages/core/src/utils/contract-verifier/helper/__tests__/getContractCreationCode.test.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { Etherscan } from '@nomicfoundation/hardhat-verify/etherscan'
+import { describe, expect, test } from 'vitest'
+import { getContractCreationCode } from '../getContractCreationCode'
+
+describe('getCreationCodeFromBlockExplorer', () => {
+  const apiKey = 'U19FHTSVAE2MTBTG89IWKF6ERWW9VX9IY4'
```
Comment (mewpawee) [inline packages/core/src/utils/contract-verifier/helper/__tests__/getContractCreationCode.test.ts]: env variable

File: packages/core/src/utils/contract-verifier/helper/__tests__/getCreationCodeWithOutMetadata.test.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  const artifacts = new ArtifactsFromPath(
+    path.join(process.cwd(), 'packages/core/src/utils/contract-verifier/helper/__mocks__/mock-artifacts'),
+  )
+  const metadata = 'a2646970667358221220e9046d113f58c010ee9536b98f2e1599c212e383de39c9cf1d7f905cc96af67c64736f6c634300080a0033'
+
+  test('with metadata', async () => {
+    const artifact = await artifacts.readArtifact('L2Pool')
+    const creationCode = await getCreationCodeWithOutMetadata(artifact.bytecode, artifact.deployedBytecode)
+    expect(creationCode).not.include(metadata)
+  })
```
Comment (tapakornl) [inline packages/core/src/utils/contract-verifier/helper/__tests__/getCreationCodeWithOutMetadata.test.ts:17]: test `with metadata` but this test check creationCode not include metadata?

File: packages/core/src/utils/contract-verifier/helper/__tests__/getCreationCodeWithOutMetadata.test.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import path from 'path'
+import { describe, expect, test } from 'vitest'
+
+import { getCreationCodeWithOutMetadata } from '../getCreationCodeWithOutMetadata'
```
Comment (tapakornl) [inline packages/core/src/utils/contract-verifier/helper/__tests__/getCreationCodeWithOutMetadata.test.ts:4]: import absolute path

File: packages/core/src/utils/contract-verifier/helper/__tests__/resolveLinkedLibraries.test.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
+import path from 'path'
+import { describe, expect, test } from 'vitest'
+
+import { createPublicClient, http } from 'viem'
+import { arbitrumSepolia } from 'viem/chains'
+
+import { getContractCreationCode } from '../getContractCreationCode'
+import { getContractInformation } from '../getContractInformation'
+import { getCreationCodeWithOutMetadata } from '../getCreationCodeWithOutMetadata'
+import { resolveLinkedLibraries } from '../resolveLinkedLibraries'
```
Comment (tapakornl) [inline packages/core/src/utils/contract-verifier/helper/__tests__/resolveLinkedLibraries.test.ts:10]: use absolute path

File: packages/core/src/utils/contract-verifier/helper/getConstructorArgsFromCreationCode.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import { Address } from 'viem'
+
+import { getContractCreationCode } from './getContractCreationCode'
+import { getCreationCodeWithOutMetadata } from './getCreationCodeWithOutMetadata'
```
Comment (tapakornl) [inline packages/core/src/utils/contract-verifier/helper/getConstructorArgsFromCreationCode.ts:4]: import abs path

**Review by mewpawee — APPROVED** [review]

File: packages/core/src/utils/contract-verifier/helper/__tests__/verifyContract.test.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
+import path from 'path'
+import { beforeAll, describe, expect, test, vi } from 'vitest'
+
+import { PublicClient, createPublicClient, http } from 'viem'
+import { arbitrumSepolia, holesky } from 'viem/chains'
+
+import { verifyContract } from '../verifyContract'
```
Comment (tapakornl) [inline packages/core/src/utils/contract-verifier/helper/__tests__/verifyContract.test.ts:7]: should we use abs path?

**Pull request merged**